### PR TITLE
OpenStack logging tools

### PR DESCRIPTION
**Pull request by exarkun**
Fixes https://clusterhq.atlassian.net/browse/FLOC-1838

Depends on #1388

[Relative diff](https://github.com/ClusterHQ/flocker/compare/blockdevice-factory-FLOC-1836...openstack-logging-FLOC-1838)

### CODE REVIEW

File: flocker/common/test/test_openstack.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                u"url": result.url,
+                u"method": result.method,
+            },
+        )
+
+    @capture_logging(lambda self, logger: None)
+    def test_keystone_client_exception(self, logger):
+        """
+        ``keystoneclient.openstack.common.apiclient.exceptions.BadRequest`` is
+        treated similarly to ``novaclient.exceptions.ClientException``.
```
Comment (tomprince) [inline flocker/common/test/test_openstack.py]: Either refer explicitly to the test whose behaviour this test is going analogous to, or explicitly explain the expected behaviour; it isn't clear that people will be reading the test docstrings in order.

File: flocker/common/_thread.py [hunk 2]
```diff
@@ -9,14 +9,46 @@
 from twisted.internet.threads import deferToThreadPool
 
 
-def _threaded_method(sync_name, method_name, reactor_name, threadpool_name):
+# TODO: Add tests and documentation for this, make it public (somewhere else).
```
Comment (tomprince) [inline flocker/common/_thread.py]: Link to an issue.

**Comment by tomprince** [comment]
After #1388 is merged, the above two minor comments are addressed, and buildbot looks statisfactory, merge.

**Comment by tomprince** [comment]
Actually, looking at where this is used, is there reason `auto_openstack_logging` doesn't just return a class that does the appropriate thing? I don't see any obvious use for the decorated class to be anything other than a class with a single field and nothing else.

**Comment by exarkun** [comment]
> Actually, looking at where this is used, is there reason auto_openstack_logging doesn't just return a class that does the appropriate thing? I don't see any obvious use for the decorated class to be anything other than a class with a single field and nothing else.

I think you're right.  I don't want to spend time refactoring this now, though.  The awkward interface is worth being able to re-use the guts of the implementation for now.  I'd be happy to see this improved later, though.

**Comment by tomprince** [comment]
You could just do
```python
def auto_openstack_logging(cls, interface):
    @_interface_decorator(...)
    class Wrapper(PRecord):
        _original = field(mandatory=true)
    return wrapper
```
or something slightly more complicated with `type` to get nicer names. That would allow you to re-use the guts without major refactoring.

On further thought, I guess maybe every use `_interface_decorator` could be refactored the same way, so maybe it doesn't make sense to do here.

**Pull request merged**